### PR TITLE
Add nav_msgs/Particle and nav_msgs/ParticleCloud messages

### PR DESCRIPTION
**Pull request by naiveHobo**
Signed-off-by: Sarthak Mittal <sarthakmittal2608@gmail.com>

As per https://github.com/ros2/rviz/issues/538 and https://github.com/ros-planning/navigation2/issues/1623, I propose adding ```nav_msgs/Particle``` and ```nav_msgs/ParticleCloud``` messages to represent particles/particle clouds produced by particle filters.

### CODE REVIEW

File: nav_msgs/msg/Particle.msg [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# This represents an individual particle with weight produced by a particle filter
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav_msgs/msg/Particle.msg:1]: > Same if you want it to be a particle that can be used with any robot navigation particle filter

I'm not bringing this up to try to argue your point, see below I think we have some philosophical overlap on naming, but I'd like to temper expectations on what this package is. `nav_msgs`, to the best of my understanding, has never been a collection of messages suitable for every form of robot navigation. If that were the case, about 2/3 of these messages shouldn't be here. These messages are here to support the work happening in the navigation stack, which often can lend themselves useful to other navigation projects. 

Anyhow.

I think your name suggestions are really good - that brings up another side of your argument that I can agree with. This should be more specific as to the type of particle it is. Personally I think the `weighted` bit is redundant. In the general formulation of a particle filter, each particle has a score so I think that's totally implied. `PoseParticle` and `PoseParticleCloud` I could totally support. 

Other examples in the community I found (still MCLs admittedly) which could readily use this definition of the messages proposed, under whatever name we decide on:
- NDT-MCL uses a PoseParticle  
- Quick-MCL uses just a visualization_msgs/MarkerArray - no weights
- mcl pi gazebo uses the PoseArray that we're trying to move from to show particle weights
- MRPT localization uses the PoseArray that we're trying to move from to show particle weights
- AMCL3D fails to publish anything at all

You're not being harsh, we're having a conversation - a good conversation :-)

[the remark this answers is not in the document]
Comment (wjwwood) [reply nav_msgs/msg/Particle.msg:1]: Ok, with the presumption that `Particle` means "state plus a single weight", then `PoseParticle*` sounds good to me.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav_msgs/msg/Particle.msg:1]: @naiveHobo are you OK with that?

[the remark this answers is not in the document]
Comment (wjwwood) [reply nav_msgs/msg/Particle.msg:1]: > nav_msgs, to the best of my understanding, has never been a collection of messages suitable for every form of robot navigation. If that were the case, about 2/3 of these messages shouldn't be here.

I agree with you, but moving forward, if I'm pressed to say "what should be in common interfaces or not" and "what should be in rviz default plugins or not", then I think it should either be "the message is generally useful for running a ROS system" or "the message is general enough to be used by multiple parties for the same purpose". Many of the messages here are here due to historical reasons.

We probably need a "generic nav messages package" and a "ROS nav(2) project specific messages package" to separate the use cases. Like many things in ROS the nav stack's messages were intended to be the standard, but it fell short.

[the remark this answers is not in the document]
Comment (naiveHobo) [reply nav_msgs/msg/Particle.msg:1]: `PoseParticle` works. I'll go ahead with the changes then?

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav_msgs/msg/Particle.msg:1]: > We probably need a "generic nav messages package" and a "ROS nav(2) project specific messages package" to separate the use cases. Like many things in ROS the nav stack's messages were intended to be the standard, but it fell short.

We sort of have that, we have nav2_msgs that are messages in a staging ground before eventually being homologated with `nav_msgs`. Whether that's `nav2_msgs` go into `nav_msgs` or `nav_msgs` gets relocated to `ros-perception`. The other things in nav2_msgs are still changing and don't belong here (yet). I suspect the actions will never belong here unless we move to `ros-perception` out of `common_interfaces` to signal the more specific project nature of it. I'm not pushing for that right now or in the near term future.

In any case, hopefully over the next few years of navigation work, we'll start being able to generalize things better. I'm slowly breaking up alot of the early assumptions and making things more flexible, but it takes time. 

@naiveHobo, yes please make those updates.

[the remark this answers is not in the document]
Comment (jacobperron) [reply nav_msgs/msg/Particle.msg]: > These messages are here to support the work happening in the navigation stack

I've never thought about the `nav_msgs` package this way before. Though indeed it is written on the [wiki](http://wiki.ros.org/nav_msgs) and in the README of this package that they are for *the* navigation stacks. Personally, I've used them in several other projects that have not used any of the other elements of the nav stack (ROS 1). It might be worth considering removing the emphasis on any particular navigation stack.

+1 for renaming to `PoseParticle` and `PoseParticleCloud`.

I think it would go a long way to get buy-in from maintainers of the other packages that produce similar data.
We can also elicit general feedback from the community with a Discourse post.

**Comment by SteveMacenski** [comment]
@wjwwood @jacobperron any other blocking reasons not to approve this now?

**Comment by SteveMacenski** [comment]
@wjwwood @jacobperron can you please take a look at this again or approve it?

**Comment by SteveMacenski** [comment]
No, the `Pose` is necessary. I don't think Tully's comment was that the `PoseParticle` was bad, I think he just wanted more context in the message definition itself on the uses of it

**Comment by sloretz** [comment]
@tfoote friendly ping :) looks like there have been changes since your last review

**Comment by tfoote** [comment]
> On units - weights are unitless normalized quantities in a particle filter s.t. sum(weights) = constant, by convention set to 1, because why chose anything else. Decoupling the RVIZ2 implementation from this PR, we could change the arrows to be representated by a color range 0-1. In practice, most of that range would never be utilized because you never have just 1 or 2 particles, usually its hundreds.

This is exactly my point. To be in a standard message these conventions should be standardized. If you use the convention of summing to 1 and I use the convention of summing to 100 our data is incompatible.  I'm not saying we can't have the datatype, but we just need to make sure it's able to stand on it's own and consider all the use cases we think are important not just know that it works for one use case and choose it. After something is accepted into common_interfaces generally it should be considered stable and we shouldn't be considering revising it or iterating on it any time soon.

> Decoupling the RVIZ2 implementation from this PR, we could change the arrows to be representated by a color range 0-1. In practice, most of that range would never be utilized because you never have just 1 or 2 particles, usually its hundreds.

This would suggest that the rviz plugin could have some different auto-scaling policies.

**Comment by SteveMacenski** [comment]
@tfoote you're asking that a ROS standard message should be used to standardize the broader field of MCLs, which is not our role to do so. We support the float64 weights which can sum to any random value and as far as the message cares, it doesn't matter. Weights do not have any absolute meaning, their all relative meanings to their distributions. 

The other discussions are topics for the rviz plugin which are reasonable. We could definitely do that without much effort.

**Comment by tfoote** [comment]
> @tfoote you're asking that a ROS standard message should be used to standardize the broader field of MCLs, which is not our role to do so. 

Yes, I'm reviewing these messages not for your application but for what a user would find if they browse through nav_msgs as part of common_interfaces, they should have value on their own or if just echo'd on the command line. The proposed messages do not refer to AMCL, generic MCL or anything else. Those use cases can be used to guide the design and make sure that there's coverage for those use cases. 

>  We support the float64 weights which can sum to any random value and as far as the message cares, it doesn't matter. Weights do not have any absolute meaning, their all relative meanings to their distributions.

With that statement they should never be communicated separately. And if I turn that around and say that since they don't have any absolute meaning in your use case, it should be fine to simply normalize them before publishing them. Thus it will still work for your use case and also provide the ability to use this message for other use cases. In particular if you have two clouds you could meaningfully fuse them if they have normalized weights.

Alternatively you could add to the message information about the distribution to which the weight should be compared to keep it as a self contained message. However if you're sending a large array of these messages, embedding the distribution information into each component significantly inflates the size of the message. Whereas adding the semantic that the weights are normalized across the distribution embeds that information into the type and doesn't require repeating it in the data fields. You would want to add the distribution information to each particle only if it was expected to vary between particle.

**Comment by SteveMacenski** [comment]
So where do we go from here? Its starting to seem like a DOA request. I see a suggestion for adding the total summed weight to the `Cloud` message so that the total weight summation is explicitly mentioned, but there are other issues you've brought up that I don't think we have a suggestion that fully satisfies your concerns. Open questions that I have seen
- Weight semantic meaning
- Standard for weight meaning
- MCL not being mentioned in a `PoseCloud` message
- Having a stand alone Pose
- Additional data that may or may not be available for a large number of filter types

That's basically every line we've suggested. I don't like missing the forest for the trees so I suppose there's a more pertinent question: Thematically, do you see some version of this being accepted? If not, then we can discuss the specifics all we want to death, but doesn't change the position that you don't think this is thematically appropriate. I'd rather be told that than death by committee where we both spend a bunch of time and energy on it. I'd really like to not revert and remove important capability from the navigation2 stack, but we're reaching that part of the road (since I don't want to deal with the influx of tickets from users saying thta they can't visualize the particle cloud when we remove support for `PoseArray`).

**Comment by jacobperron** [comment]
> do you see some version of this being accepted?

IMO, some version of the particle cloud message could live in `nav_msgs`.

>  After something is accepted into common_interfaces generally it should be considered stable and we shouldn't be considering revising it or iterating on it any time soon.

With this is mind, it might be better for the proposed messages to live in a different package outside of `common_interfaces` (i.e. `nav2_msgs`, where they currently live) until they are more mature and then move them here at a later date. I don't know if there is precedence for something like this, but it seems reasonable to me. From my scans of the related tickets, it sounds like the main intention is to use the new weights to improve tooling (e.g. visualization and analytics). It might be good to wait and see if the development of such tooling leads to changes of the initial message definitions.

I'm okay to continue iterating in this PR and landing the message in `common_interfaces`, if other team members are on-board.

**Comment by SteveMacenski** [comment]
That's helpful, thanks. I suppose it might be helpful if we were to bring up problems and potential tangible solutions to those problems. I'm all for more docs, more clarity, and either attempting to be more general or more specific. I think I need some help in how to turn the issues being brought up into progress. 

As to your second comment, we do have the message in nav2_msgs that we can update with the things we did here as well (and the rviz plugin to do autoscaling). I don't see as part of that work any other necessary changes, expect adding a "total_sum" field to avoid iterating over each particle to sum values for scale. So if there's other suggestions of changes, docs or specifications to add, etc, I think the discussion here more or less continues per usual :-)

**Comment by SteveMacenski** [comment]
Pinging on this - I'd **really** like to have this continue to move forward. Its blocking non-trivial improvements across the ecosystem.

**Comment by mikeferguson** [comment]
As one of the maintainers of ROS1 Navigation for the past 7 years, as well as someone who has developed several commercial mobile robots, I figured I would pipe in when I saw this.

The messages in nav_msgs are not generally generic enough for every use case: they mainly target 2d navigation (which, is a pretty big use case). For instance, all of the map messages are only good for 2d. This message pair seems to be at least as generic as that (hey, it even supports 3d!) - and if we consider primarily 2d navigation, most robots are running some variant of an MCL particle filter for localization. That particle filter generally consists of a set of weighted particles.

> I can tell you from experience that those weights are really important for robust delocalization detection of particle filter based localization approaches. It is not just for visualization. Really, that's what I'm trying to enable here because that's been a common complaint of the nav stack in ROS1 that there's insufficient information to do even remotely reasonable delocalization detection.

I agree here. I've seen that change done several times. It was never top priority to get it merged back into the open source, hence why ROS1 doesn't support it.

I would also agree with some above that the Particle message doesn't really stand on it's own - but without it, the ParticleCloud message would be quite nasty (unless there is some ROS2 feature I'm not aware of that let's us define Particle inside the ParticleCloud message).

So, in conclusion, is this the absolute most important feature to ever get into ROS2? No, but I think it would be a pretty good improvement towards standardizing the messages used in ROS2 localization systems.

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/pull/1677#issuecomment-682424964

There's a clear pattern that companies are hacking messages to get particle weight information due to a lack of a standard message. This now represents another company that have had to make modifications due to a lack of a standard message.

**Comment by tfoote** [comment]
I do think that a version of these messages can live in nav_msgs, and I would like to see that. I've identified areas where the proposed message is not clearly defined. 

The message must be meaningful and documented fully so that if I come across the message definition I can interpret every single component of the message in the same way that everyone else is. 


> Weight semantic meaning
> Standard for weight meaning

If an apple is labeled as 23 without units. You can only get value out of that label by guessing it's unit. If we just agree that the weights of points in the particle cloud sum to 100% it's very easy to interpret the weight of a single sample when processing. And you don't have to sum up all the other particles in the cloud to have a sense of what the weight represents. 

The weights effectively are forming a discrete weight function to influence the results of a computation based on the inputs. This is just a semantic meaning that needs to be documented. But while were there we can normalize the weights like we do for most other 

This is the most fundamental part of a message definition It needs to describe what this field means, so that an implementer can know how to interpret the values of a given weight. What does a weight of 10 mean? It would be trivial to put into the comments that with weights are the relative importance of any particle on the range 0.0 to 1.0 and sum to 1.0 in a ParticleCloud. This gives you a clear meaning for the meaning of the weight on any given particle, it's self contained and weights have consistent deterministic meaning for a specific particle that can then even be compared across instances of clouds.

> MCL not being mentioned in a PoseCloud message
> Having a stand alone Pose
> Additional data that may or may not be available for a large number of filter types

We should define what use cases we want to cover and which ones we don't want to cover. If we make a super generic "standard" message that doesn't actually cover any use cases then we've defeated the goal of being useful for code reuse. This proposal doesn't call out the specific use cases to cover and thus it's very hard to evaluate the suitability of the message to cover the desired use case(s) for their use. If we're all imagining different use cases in our discussions then we're likely going to talk past each other. Once we understand what the use cases specifically are then the messages can be evaluated as to whether they're meeting them.

I'm coming to this as a non-expert in the navigation field. That's for whom these messages need to be documented so that there's no ambiguity that needs an expert. I hear that there's pressure to do this and people are forging their own way. That's actually a really good way to collect the use cases, find those collect them here and then make sure that we're covering what they need. I don't think that these messages are very far off. But if we're going to stamp them as standard I do really want to make sure that they're actually standard. We made PointCloud and then made PointCloud2 a few months later. And it took us close to 10 years to retired PointCloud. I don't want to do that again.

**Comment by SteveMacenski** [comment]
I think this is alot easier if we talk in specifics. 

`PoseParticle` - as of now

```
# This represents an individual particle with weight produced by a particle filter
geometry_msgs/Pose pose
float64 weight
```

`PoseParticleCloud` - as of now
```
# This represents a particle cloud containing particle poses and weights
std_msgs/Header header
# Array of particles in the cloud
PoseParticle[] particles
```

---

`PoseParticle` - Proposed to resolve concerns

```
# This represents an individual particle with weight produced by a particle filter
geometry_msgs/Pose pose # pose of the particle in the PoseParticleCloud
float64 weight # unitless weight of a particle in the PoseParticleCloud
```

`PoseParticleCloud` - Proposed to resolve concerns
```
# This represents a particle cloud containing particle poses and weights
std_msgs/Header header
# Array of particles in the cloud
PoseParticle[] particles # an array of particles in the cloud
Float64 particle_weight_sum # total sum of all particle's weights
```

This annotates all the values to tell people exactly what's going on and mentions units on weights and the total sum of the weights to be used as appropriate. By default, a reasonable thing to do would be to have the weights sum to 1, but that will not always be the case. It may be reasonable to consider each particle as a pdf containing itself a value between 0-1 and simply scaling them isn't appropriate for the application (or really matters at all). 

Is there any modifications you would like to make or request in addition? I think the "Pose" in front of particle / cloud makes it clear what the type of particle is; could be full SE3, SE2, or 2D. The weights are described as unitless, as they are for most every particle filter variants I'm aware of, with a total sum to scale them to a normalized value if necessary.

**Comment by hidmic** [comment]
@tfoote friendly ping

**Comment by SteveMacenski** [comment]
@tfoote nudge from 2021 :-)
